### PR TITLE
feat: add license contract

### PR DESCRIPTION
**Pull request by alaibe**

### CODE REVIEW

File: contracts/license.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    uint256 price;
+
+    mapping(address => uint256) licenseOwners;
+
+    constructor(address payable _recipient, uint256 _price) public {
+        recipient = _recipient;
+        price = _price;
+    }
+
+    function isLicenseOwners() public view returns (bool) {
```
Comment (richard-ramos) [inline contracts/license.sol]: It would be a good idea to receive the address to verify if it's a licensed owner, in case this contract is called from another contract

File: contracts/license.sol [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        return licenseOwners[msg.sender] != 0;
+    }
+
+    function buy() public payable {
+        require(msg.value == price, "Value is not equal to expected price");
+        require(licenseOwners[msg.sender] == 0, "License already bought");
+
+        licenseOwners[msg.sender] = msg.value;
+        recipient.transfer(msg.value);
```
Comment (richard-ramos) [inline contracts/license.sol]: It's a good idea for functions that affect state, to emit events.

File: contracts/license.sol [hunk 3]
```diff
@@ -0,0 +1,43 @@
+pragma solidity ^0.5.0;
+
+import "./ownable.sol";
+
```
Comment (richard-ramos) [inline contracts/license.sol]: Contract/Functions are missing natspec (fancy word for solidity comments :) )

File: contracts/license.sol [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+contract License is Ownable {
+    address payable recipient;
```
Comment (richard-ramos) [inline contracts/license.sol]: Add access modifier (public/private/etc)

File: app/js/layout/App.jsx [hunk 5]
```diff
@@ -1,27 +1,59 @@
...
+    this.state = { ready: false }
+  }
+
+  componentDidMount() {
+    this.props.fetchPrices(relevantPairs);
+
+    EmbarkJS.onReady(async (err) => {
+      if (!err) {
+        this.setState({ ready: true })
+      }
```
Comment (alaibe) [inline app/js/layout/App.jsx]: Make sure we don't make request before embarkjs is ready.
I feel we can move that into global state in another PR

Agree?

File: app/js/index.js [hunk 6]
```diff
@@ -8,9 +8,6 @@ import 'bootstrap/dist/css/bootstrap.css';
 import App from './layout/App';
 import history from './history';
 import store from './store';
-import init from './init';
-
-init();
```
Comment (alaibe) [inline app/js/index.js]: moved to app.jsx so we can manage the display

File: contracts/license.sol [hunk 7]
```diff
@@ -0,0 +1,82 @@
+pragma solidity ^0.5.0;
+
+import "./ownable.sol";
+
+/**
+* @title License
+* @dev License contract for buying a license
+*/
+contract License is Ownable {
```
Comment (bgits) [inline contracts/license.sol]: Is there some kind of limitation on the number of licenses? Should there be a way for a licenseOwner to revoke their license and get the fee back?

File: app/js/layout/App.jsx [hunk 8]
```diff
@@ -1,27 +1,59 @@
...
+};
 
 class App extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { ready: false }
+  }
+
+  componentDidMount() {
+    this.props.fetchPrices(relevantPairs);
```
[the remark this answers is not in the document]
Comment (alaibe) [reply app/js/layout/App.jsx]: Updated and moved to constructor.

File: contracts/license.sol [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        return licenseOwners[_address] != 0;
+    }
+
+    /**
+    * @dev Buy a license
+    * @notice Requires value to be equal to the price of the license.
+    *         The buyers must not already own a license.
+    */
+    function buy() public payable {
+        require(msg.value == price, "Value is not equal to expected price");
```
[the remark this answers is not in the document]
Comment (alaibe) [reply contracts/license.sol]: at this stage, yes

File: contracts/license.sol (hunk 7 above)
Comment (alaibe) [reply contracts/license.sol]: This is only about buying a license

File: app/js/layout/App.jsx (hunk 5 above)
Comment (jrainville) [reply app/js/layout/App.jsx]: Yeah we should. We a loader in the meantime

File: test/license_spec.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    const recipient = await License.methods.getRecipient().call();
+    const price = await License.methods.getPrice().call();
+    assert.strictEqual(parseInt(price, 10), 1);
+    assert.strictEqual(recipient, zeroAddress);
+  });
+
+
+  it("should not allow to buy license when price is incorrect", async function () {
+    try {
+      await License.methods.buy().send({value: 2, from: accounts[0]});
```
Comment (jrainville) [inline test/license_spec.js]: To make the test more "accurate", you could test for 1.1 or 0.9

Comment (jrainville) [reply test/license_spec.js]: Unless the price is in Wei

File: test/license_spec.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      assert.strictEqual(error.message, "VM Exception while processing transaction: revert License already bought");
+    }
+  });
+
+  it("should allow set the price if owner", async function () {
+    try {
+      await License.methods.setPrice(10).send({from: accounts[1]});
+    } catch(error) {
+      assert.strictEqual(error.message, "VM Exception while processing transaction: revert");
+    }
```
Comment (jrainville) [inline test/license_spec.js]: Could you split this test case in 2. One for it should and one for it should not.

File: test/license_spec.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    const price = await License.methods.getPrice().call();
+    assert.strictEqual(parseInt(price, 10), 10);
+  });
+
+  it("should allow set the recipient if owner", async function () {
+    try {
+      await License.methods.setRecipient(accounts[1]).send({from: accounts[1]});
+    } catch(error) {
+      assert.strictEqual(error.message, "VM Exception while processing transaction: revert");
+    }
```
Comment (jrainville) [inline test/license_spec.js]: Same here. Thanks
